ouroboros-network-protocols/bench-cddl/Main.hs

```diff
@@ -0,0 +1,439 @@
+                        (MsgSharePeers [ SockAddrInet maxBound addr
+                                       | x <- [0..maxBound]
+                                       , let addr = tupleToHostAddress (192, 168, 1, x)
+                                       ])
+  ]
+
+myBenchCodec :: ( forall (st :: ps) (st' :: ps). NFData (Message ps st st')
+                , forall (st :: ps) pr. NFData (PeerHasAgency pr st)
+                )
+             => String
+             -> (NodeToNodeVersion -> Codec ps DeserialiseFailure IO ByteString)
+             -> NodeToNodeVersion
+             -> AnyMessageAndAgency ps
+             -> [Benchmark]
+myBenchCodec title getCodec ntnVersion msg =
+  [ myBenchEncode (title ++ " " ++ "Encode") getCodec ntnVersion msg
+  , myBenchDecode (title ++ " " ++ "Decode") getCodec ntnVersion msg
+  ]
+
+-- TODO: We don't force the agency along with the message because this leads
+-- to a weird error that could be a tasty-bench bug.
+-- See: https://github.com/Bodigrim/tasty-bench/issues/52
+--
+-- Only LocalStateQuery agency tokens carry the query information so it
+-- shouldn't be too problematic.
+myBenchEncode :: ( forall (st :: ps) (st' :: ps). NFData (Message ps st st')
+                 , forall (st :: ps) pr. NFData (PeerHasAgency pr st)
+                 )
+              => String
+              -> (NodeToNodeVersion -> Codec ps DeserialiseFailure IO ByteString)
+              -> NodeToNodeVersion
+              -> AnyMessageAndAgency ps
+              -> Benchmark
+myBenchEncode title getCodec ntnVersion (AnyMessageAndAgency a m) =
+  let Codec { encode } = getCodec ntnVersion
+   in env (pure (a, m)) $ \ ~(agency, message) ->
+        bench title $ nf (encode agency) message
+
+-- TODO: We don't force the agency along with the message because this leads
+-- to a weird error that could be a tasty-bench bug.
+-- See: https://github.com/Bodigrim/tasty-bench/issues/52
+--
+-- Only LocalStateQuery agency tokens carry the query information so it
+-- shouldn't be too problematic.
+myBenchDecode :: ( forall (st :: ps) (st' :: ps). NFData (Message ps st st')
+                 , forall (st :: ps) pr. NFData (PeerHasAgency pr st)
+                 )
+              => String
+              -> (NodeToNodeVersion -> Codec ps DeserialiseFailure IO ByteString)
+              -> NodeToNodeVersion
+              -> AnyMessageAndAgency ps
+              -> Benchmark
+myBenchDecode title getCodec ntnVersion (AnyMessageAndAgency a m) =
+  let Codec { encode
+            , decode
+            } = getCodec ntnVersion
+   in env (pure (a, encode a m)) $ \ ~(agency, encodedMessage) ->
+        bench title $ nfIO (do
+                              decoder <- decode agency
+                              res <- runDecoder [encodedMessage] decoder
+                              return $ case res of
+                                Left err -> Just err
+                                Right {} -> Nothing
```

Comment:
Shouldn't it be a strict pattern match? From `Right` we know we successfully decoded the bytestring, so what the bang really adds is just evaluating the newly constructed Haskell value.
```suggestion
                                Right !_ -> Nothing
```